pilz_control/include/pilz_control/traj_mode_manager.h

```diff
@@ -62,65 +62,57 @@ class TrajProcessingModeStateMachine
  */
-class TrajProcessingModeListener
+class HoldModeListener
 {
 public:
-  //! @param mode Mode to wait for.
-  TrajProcessingModeListener(const TrajProcessingMode& mode);
-  void waitForMode();
-  //! @returns true if the given mode corresponds to the target mode, otherwise false.
-  bool isTargetMode(const TrajProcessingMode& mode) const;
-  //! @brief Notify the listener that the target mode is reached.
+  void waitForHold();
+  //! @brief Notify the listener that the hold mode is reached.
   void triggerListener();
 
 private:
   std::mutex mutex_;
   std::condition_variable cond_variable_;
   bool cond_fulfilled_ {false};
-  //! @brief Mode to wait for.
-  const TrajProcessingMode mode_;
 };
 
 /**
  * @brief Encapsulates a state machine managing the current Trajectory-Processing-Mode.
+ * 
+ * All events are mutually exclusive. Initial mode is stopping in accordance to the controller.
  */
 class TrajProcessingModeManager
 {
 public:
-  //! @returns true only if a successful state switch to stopping was performed, otherwise false.
-  bool stoppingEvent();
+  //! @return True only if a successful state switch to stopping was performed, otherwise false.
+  bool stopEvent(HoldModeListener* const listener = nullptr);
   //! @brief Switch to hold.
   void stopMotionFinishedEvent();
-  //! @returns true if in state unhold or a successful switch to state unhold was performed, otherwise false.
-  bool unholdEvent();
+  //! @return True if in state unhold or a successful switch to state unhold was performed, otherwise false.
+  bool startEvent();
 
 public:
   //! @brief Check if in state stopping or hold.
   bool isHolding();
-  bool isUnhold();
-  void registerListener(TrajProcessingModeListener* const listener);
-
   TrajProcessingMode getCurrentMode();
 
 private:
-  //! @brief Perform full transition if possible.
-  bool switchTo(const TrajProcessingMode& mode, const bool success_at_transition_only=true);
-  bool setMode(const TrajProcessingMode& requested_mode, const bool& success_at_transition_only);
+  /**
+   * @brief Perform transition if possible.
+   * @return True if transition was performed, otherwise false.
+   */
+  bool switchTo(const TrajProcessingMode& mode);
+  void registerHoldListener(HoldModeListener* const listener);
   //! @brief Triggers all registered listeners whose target mode is reached.
-  void callListener(const TrajProcessingMode& mode);
-  //! @brief Use only if lock on \ref mode_mutex_ is already acquired.
+  void callHoldListener();
 
 private:
   const TrajProcessingModeStateMachine mode_state_machine_;
   TrajProcessingMode current_mode_ {TrajProcessingMode::stopping};
-  //! @brief Protects the access to the current mode.
-  std::mutex mode_mutex_;
-
-  std::list<TrajProcessingModeListener*> listener_;
-  //! @brief Protects the access to the listener list.
-  std::mutex listener_mutex_;
+  std::list<HoldModeListener*> listener_;
+  //! @brief Protects the access to member variables.
```

Comment:
To general because e.g. it does not protect access to `mode_state_machine_`. I suggest specifically name the member which are protected. Or even better would be a mutex name which already indicates what it protects.

Reply:
Actually, it protects `mode_state_machine_` indirectly, because it is used to make all public methods mutually exclusive. I changed the description to make that clear.